### ISSUE TITLE
Running install hangs

### ISSUE THREAD

**Issue by adamciolkowski** [pnpm/pnpm#946]
pnpm hangs when running a clean install. It also happens when I try to install a single, relatively large package (i. e. webpack or react).

After downloading some of the packages fairly quickly the command just hangs without any progress:
```
$ pnpm install webpack-dev-server@2.7.1
$ Resolving: total 250, reused 1, downloaded 242
```
I tried to limit the number of HTTP connections. It results in the hang happening much earlier:
```
$ pnpm install webpack-dev-server@2.7.1 --network-concurrency 1
$  WARN  waiting for another installation to complete...
$ Resolving: total 13, reused 0, downloaded 13
```

I tried removing contents of `~./pnpm-store`.
I'm not using a VPN.

I used:
 - pnpm version: 1.23.1
 - `node -v` prints: v6.12.0
 - Linux

**Comment by zkochan** [comment]
Does it happen all the time or from time to time?

Have you tried to wait for a minute or two? Does it eventually finish the installation?

**Comment by adamciolkowski** [comment]
I've been able to install several smaller packages with no problem, but it always hangs on packages that have a lot of dependencies themselves. In those cases I've waited up to 15 minutes during which there was no progress at all.

Is there an option to make the output more verbose? The only thing I came up with was adding a readPackage hook to log packages.

Also, does it work for you, were you able to reproduce it on your end?

**Comment by zkochan** [comment]
> Is there an option to make the output more verbose? The only thing I came up with was adding a readPackage hook to log packages.

you can run `pnpm i --reporter ndjson`

> Also, does it work for you, were you able to reproduce it on your end?

yes, it works on my linux. It does hang sometimes (when I run this benchmarks: https://github.com/pnpm/node-package-manager-benchmark) but not forever

**Comment by zkochan** [comment]
@adamciolkowski is it still reproducible? We had a few big updates in pnpm since `pnpm@1.23.1`. We removed usages of request (#953) which might have caused something like this issue

**Comment by adamciolkowski** [comment]
Sorry for the late response. I've been busy and hadn't had time to check the newer version of pnpm yet. I'll post an update when I do.

**Comment by tomcon** [comment]
I am getting hanging all the time with no way around this, can someone help please?
I've tried updating, uninstalling, re-installing but zero joy
Is there a way of manually forcing an uninstall??

pnpm list -g --depth=0
pnpm-global-pkg@1.0.0 C:\Bin\Node\pnpm-global\1
├── ...
├── svelte@2.5.1
├── svelte-cli@3.0.0
├── svelte-extras@2.0.2
├── svelte-loader@2.9.0
├── svelte-transitions@1.1.1
└── svelte-upgrade@1.0.16

pnpm -g uninstall svelte
 WARN  waiting for another installation to complete...
Packages: -1
-
^CTerminate batch job (Y/N)? y

pnpm -g install svelte
 WARN  waiting for another installation to complete...
Resolving: total 1, reused 0, downloaded 1

**Comment by lotri1** [comment]
I'm also experiencing hangs when pnpm downloads packages.
If packages are reused, it won't hang.
I'm using pnpm version 2.0.0.

**Comment by zkochan** [comment]
@lotri1 are the packages that hang big? See #1131

**Comment by lotri1** [comment]
It looked to me like a deadlock between threads that handling the packages.

Anyway, I couldn't find a solution so I moved to npm ci.

**Comment by zkochan** [comment]
What OS did you use?

**Comment by abrsil-WittGruppe** [comment]
I evaluated pnpm for a few days and ran into this issue a couple of times.

```
$ pnpm -v
2.12.2

$ uname -a
MINGW64_NT-10.0 AZEESS 2.9.0(0.318/5/3) 2017-10-05 15:05 x86_64 Msys
```

I tried reproducing it but it seems to be random.
After some time, install starts working again.

**Comment by n4bb12** [comment]
I evaluated pnpm for a few days and ran into this issue a couple of times.

```
$ pnpm -v
2.12.2

$ uname -a
MINGW64_NT-10.0 AZEESS 2.9.0(0.318/5/3) 2017-10-05 15:05 x86_64 Msys
```

I tried reproducing it but it seems to be random.
After some time, install starts working again.

**Comment by n4bb12** [comment]
Okay seems to happen when I hit `CTRL+C` during an install.
I searched [fs-locker](https://github.com/pnpm/fs-locker) but it doesn't seem to handle signals.
So my theory is that pnpm gets interrupted while something is locked and the lock remains until after a while it becomes stale.

**Comment by zkochan** [comment]
I have a found a place where pnpm can silently hang: https://github.com/pnpm/package-store/blob/master/src/storeController/index.ts#L45-L48

When pnpm waits for a store lock, nothing is printed to the console.

**Comment by zkochan** [comment]
I fixed the store reporting and some locking issues in pnpm v2.13.1

Let me know if this will repeat in v2.13.1 or newer. Also, include the output (and the `pnpm-debug.log`, if present)

**Comment by zkochan** [comment]
Let's create a new issue in case it will happen again with latest pnpm

**Comment by patricksile** [comment]
Hi, I am having this hanging issue also when running `pnpm i -g webpack-dev-server` , it happens with large modules actually. I waited for more than 10 mins and nothing happened.

**Comment by zkochan** [comment]
@patricksile file an issue with more details
os, node version, etc

**Comment by patricksile** [comment]
I am using Windows 10 Home October Update with Ubuntu 18.04 LTS in WSL (Windows Subsystem Linex), `node v11.3.0` and `pnpm 2.19.4`. 
And the issue is the following:
- When I run `pnpm i -g webpack-dev-server` and some other apparently large modules the installation process just hangs.

**Comment by dandv** [comment]
Same issue with 2.23.1 on Linux. To reproduce, run `pnpm multi install` and press Ctrl+S to pause the process. Wait a few minutes, then press Ctrl+Q to resume.

```
$ pnpm multi install
Scope: all 3 workspace packages
Resolving: total 33, reused 0, downloaded 24
/usr/lib/node_modules/pnpm/lib/node_modules/proper-lockfile/index.js:186
    compromised = compromised || function (err) { throw err; };
                                                  ^

Error: Unable to update lock within the stale threshold
    at options.fs.utimes (/usr/lib/node_modules/pnpm/lib/node_modules/proper-lockfile/index.js:108:66)
    at FSReqCallback.args [as oncomplete] (fs.js:147:20)

$ pnpm multi install
Scope: all 3 workspace packages
.                                        |  WARN  waiting for another installation to complete...
```

^ never finishes

**Comment by Apu-Emdad** [comment]
command: `pnpm install`
**stuck at:**

> Lockfile is up to date, resolution step is skipped
> Packages: +1252
> ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
> Downloading registry.npmjs.org/typescript/5.1.6: 7.15 MB/7.15 MB, done
> Packages are hard linked from the content-addressable store to the virtual store.
>   Content-addressable store is at: D:\.pnpm-store\v3
>   Virtual store is at:             node_modules/.pnpm
> Progress: resolved 1252, reused 1251, downloaded 1, added 1

node version :  v18.17.0
pnpm version: 8.6.9

**Comment by JustMrMendez** [comment]
> 6 years later... Insane
> 
> `pnpm install` hangs indefintitely
> 
> OS: MacOS Ventura Arch: Apple M2 Max NodeVersion: 18.16.0
> 
> ```shell
> ┬─[thebeachmaster@orion:~/g/t/???]─[17:00:25]─[G:develop=]
> ╰─>$ pnpm install                                                                                                                                                                                                     164ms  Thu Aug 31 17:00:25 2023
> Downloading registry.npmjs.org/next/13.4.19: 12.97 MB/12.97 MB, done
> Downloading registry.npmjs.org/react-icons/4.10.1: 20.69 MB/20.69 MB, done
> Progress: resolved 18, reused 0, downloaded 18, added 0
> 
> 
> <--Hang for 6 minutes, I had to cancel--->
> 
> ^C⏎                                                                                                                                                                                                                                                   ┬─[thebeachmaster@orion:~/g/t/???]─[17:07:43]─[G:develop=]
> ╰─>$ npm install                                                                                                                                                                                                       6.6m  Thu Aug 31 17:07:43 2023
> 
> added 357 packages, and audited 358 packages in 37s
> 
> 109 packages are looking for funding
>   run `npm fund` for details
> 
> 3 vulnerabilities (2 moderate, 1 high)
> 
> To address all issues, run:
>   npm audit fix
> 
> Run `npm audit` for details.
> ┬─[thebeachmaster@orion:~/g/t/???]─[17:08:43]─[G:develop=]
> ╰─>$  
> ```

updated note to 20.5 and it worked

**Comment by UnsolvedCypher** [comment]
I also hit this, and upgrading to Node 20+ worked for me as well.

**Comment by synapdk** [comment]
Docker image: `node:20.9.0-alpine`
`pnpm@8.7.6`

I have a Bitbucket pipeline docker image build failing on `pnpm fetch`. It either gets `exit code: 137` or runs out of memory. This started with the `node:18.18.2-alpine` image and I was hopeful that switching to Node 20 would solve it, but it did not. 

```#10 [build 4/9] RUN pnpm fetch
#10 sha256:5adff1481f6601c1ae2350e5e1525317bd50a5ec3bdf24529736a4f0745ca518
#10 1.869 Importing packages to virtual store
#10 1.944 Already up to date
#10 2.036 Progress: resolved 1, reused 0, downloaded 0, added 0
#10 3.051 Progress: resolved 1566, reused 0, downloaded 0, added 0
#10 4.055 Progress: resolved 1566, reused 0, downloaded 13, added 4
#10 5.051 Progress: resolved 1566, reused 0, downloaded 23, added 16
#10 6.124 Progress: resolved 1566, reused 0, downloaded 127, added 124
#10 7.136 Progress: resolved 1566, reused 0, downloaded 168, added 164
#10 8.251 Progress: resolved 1566, reused 0, downloaded 208, added 200
#10 9.261 Progress: resolved 1566, reused 0, downloaded 290, added 286
#10 10.35 Progress: resolved 1566, reused 0, downloaded 370, added 364
#10 16.66 Progress: resolved 1566, reused 0, downloaded 376, added 371
#10 ERROR: executor failed running [/bin/sh -c pnpm fetch]: exit code: 137
```

**Comment by paulcarroll** [comment]
This is still an issue for me *most* times I upgrade pnpm
Using M1 Macbook Pro, Node 20.5.1 (but it has been happening for the last couple of years)

Unsure as to what fixes it, IIRC deleting the installation and reinstalling does the trick. Also, I cannot kill the hung pnpm processes through zsh or from activity monitor

**Comment by hoghweed** [comment]
I have the same issue even today. 
- I updated to node 20.16.0
- rimraf'ed the node_modules
- rebooted 

but nothing now I can't unblock and I'm stuck.....

**Comment by TheHussein** [comment]
Facing this issue when running pnpm in a docker container locally, but works in production, this seems to work if I downgrade to node 16 and pnpm 8, but that's obviously not a solution.

**Comment by TheHussein** [comment]
Downgraded from Node 22 to latest Node 20 fixed it on PNPM 9.7.0 for me.

**Comment by facundocapua** [comment]
I was having the same issue on pnpm 9.10.0 and node 20. I had to downgrade to pnpm 9.7.0 to make it work.

**Comment by zkochan** [comment]
If pnpm hangs in some scenario please create steps to reproduce for us in a new issue.
